### PR TITLE
API: Update the ssl context after each accepting incoming connection

### LINKED ISSUE #8501: CRL expires if the daemon is not reloaded/restarted
## Describe the bug
A very simple configuration of icinga2 was unable to connect to our master, because the CRL has expired. The daemon was not reloaded for 30 days, which is our maximum CRL age.
We update our CRL every 6 hours and a restart fixed the issue. So I think, that a running daemon is not reloading a changed CRL.

```
Sep 24 10:06:54 XX icinga2[4217]: API client disconnected for identity 'XXX'
Sep 24 10:06:59 XX icinga2[4217]: Certificate validation failed for endpoint 'XXX': code 12: CRL has expired
```
## To Reproduce
Start icinga2, wait until CRL expiration and connections start to fail.

## Expected behavior
The daemon periodically reloads the CRL or monitors the CRL for changes.

## Your Environment
Application version: 2.12.1-1.buster

System information:
  Platform: Debian GNU/Linux
  Platform version: 10 (buster)
  Kernel: Linux
  Kernel version: 4.19.0-12-amd64
  Architecture: x86_64

Enabled features: api checker command mainlog notification syslog

Config validation:
```
[2020-11-24 11:08:59 +0100] information/cli: Icinga application loader (version: r2.12.1-1)
[2020-11-24 11:08:59 +0100] information/cli: Loading configuration file(s).
[2020-11-24 11:08:59 +0100] information/ConfigItem: Committing config item(s).
[2020-11-24 11:08:59 +0100] information/ApiListener: My API identity: XXX
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 NotificationComponent.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 SyslogLogger.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 Host.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 7 NotificationCommands.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 FileLogger.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 2 Notifications.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 IcingaApplication.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 EventCommand.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 CheckerComponent.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 2 Zones.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 2 Endpoints.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 ExternalCommandListener.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 ApiListener.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 156 CheckCommands.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 7 TimePeriods.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 10 UserGroups.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 31 Users.
[2020-11-24 11:09:00 +0100] information/ConfigItem: Instantiated 1 Service.
[2020-11-24 11:09:00 +0100] information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
[2020-11-24 11:09:00 +0100] information/cli: Finished validating the configuration file(s).
```

## Additional context
We are rolling out our own certificate infrastructure and are not relying on icinga2 pki.

### PR DESCRIPTION
**Pull request by Yonas-net**
resolves #8501

### CODE REVIEW

File: lib/remote/apilistener.cpp [hunk 1]
```diff
@@ -432,9 +432,14 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
 
 	for (;;) {
 		try {
-			auto sslConn (Shared<AsioTlsStream>::Make(io, *sslContext));
+			asio::ip::tcp::socket socket(io);
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:435]: Please write `socket (io)`, not `socket(io)`.

File: lib/remote/apilistener.cpp [hunk 2]
```diff
@@ -432,9 +432,14 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
...
+
+			server->async_accept(socket.lowest_layer(), yc);
 
-			server->async_accept(sslConn->lowest_layer(), yc);
+			UpdateSSLContext();
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:439]: You update it on every new connection independent of whether the CRL has changed or not. Are you sure this is needed?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: lib/base/utility.cpp [hunk 3]
```diff
@@ -1310,6 +1310,14 @@ bool Utility::PathExists(const String& path)
 	return fs::exists(fs::path(path.Begin(), path.End()), ec) && !ec;
 }
 
+time_t Utility::GetFileCreationTime(const String& path)
+{
+	namespace fs = boost::filesystem;
+	fs::path filePath(path);
```
Comment (Al2Klimov) [inline lib/base/utility.cpp:1316]: Why a local variable?

File: lib/remote/apilistener.cpp [hunk 4]
```diff
@@ -429,12 +429,19 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
 	namespace asio = boost::asio;
 
 	auto& io (IoEngine::Get().GetIoContext());
+	time_t lastModified = 0;
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:432]: 0? That's not true.

File: lib/remote/apilistener.cpp [hunk 5]
```diff
@@ -429,12 +429,19 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
...
 	for (;;) {
 		try {
-			auto sslConn (Shared<AsioTlsStream>::Make(io, *sslContext));
+			asio::ip::tcp::socket socket (io);
+
+			server->async_accept(socket.lowest_layer(), yc);
 
-			server->async_accept(sslConn->lowest_layer(), yc);
+			if (lastModified != Utility::GetFileCreationTime(GetCrlPath()))
+				UpdateSSLContext();
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:441]: Use { }.

File: lib/remote/apilistener.cpp [hunk 6]
```diff
@@ -429,12 +429,19 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
...
 
 	for (;;) {
 		try {
-			auto sslConn (Shared<AsioTlsStream>::Make(io, *sslContext));
+			asio::ip::tcp::socket socket (io);
+
+			server->async_accept(socket.lowest_layer(), yc);
 
-			server->async_accept(sslConn->lowest_layer(), yc);
+			if (lastModified != Utility::GetFileCreationTime(GetCrlPath()))
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:440]: What happens if I have no CRL configured?

File: lib/remote/apilistener.cpp (hunk 5 above)
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:441]: UpdateSSLContext() disconnects all already established connections. Do we want this?

File: lib/remote/apilistener.cpp [hunk 7]
```diff
@@ -443,6 +450,8 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
 			Log(LogCritical, "ApiListener")
 				<< "Cannot accept new connection: " << ex.what();
 		}
+
+		lastModified = Utility::GetFileCreationTime(GetCrlPath());
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:454]: Why here and not after accept?

File: lib/base/utility.hpp [hunk 8]
```diff
@@ -109,6 +109,7 @@ class Utility
 	static tm LocalTime(time_t ts);
 
 	static bool PathExists(const String& path);
+	static time_t GetFileCreationTime(const String& path);
```
Comment (Al2Klimov) [inline lib/base/utility.hpp:112]: The addition of this one shall be an extra commit with a proper message.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: lib/base/utility.hpp (hunk 8 above)
Comment (Yonas-net) [reply lib/base/utility.hpp:112]: with a proper commit message in the same PR or also an own PR?

File: lib/base/utility.cpp (hunk 3 above)
Comment (Yonas-net) [reply lib/base/utility.cpp:1316]: why should I need a member variable, which is only needed in this function?

File: lib/base/utility.hpp (hunk 8 above)
Comment (Al2Klimov) [reply lib/base/utility.hpp:112]: Same PR.

File: lib/base/utility.cpp (hunk 3 above)
Comment (Al2Klimov) [reply lib/base/utility.cpp:1316]: I didn’t say anything about member variables. I just asked what's this local variable for?

Comment (Yonas-net) [reply lib/base/utility.cpp:1316]: Because the function does not accept string paths, but `boost::filesystem::path&` instead.

Comment (Al2Klimov) [reply lib/base/utility.cpp:1316]: What about casting the string to a path then?

Comment (Yonas-net) [reply lib/base/utility.cpp:1316]: Yes, that would work too!

File: lib/base/utility.cpp [hunk 9]
```diff
@@ -1310,6 +1310,13 @@ bool Utility::PathExists(const String& path)
 	return fs::exists(fs::path(path.Begin(), path.End()), ec) && !ec;
 }
 
+time_t Utility::GetFileCreationTime(const String& path)
+{
+	namespace fs = boost::filesystem;
+
+	return fs::last_write_time(static_cast<fs::path>(path));
```
Comment (Al2Klimov) [inline lib/base/utility.cpp:1317]: Why a static cast and not a function style cast?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
1. Call GetCrlPath() only once at all
2. Call Utility::GetFileCreationTime() at most once per accept.
3. If you just start Icinga 2 with an empty CRL configured, don't touch the CRL after start and establish your first connection, will Icinga 2 update the context?

File: lib/base/utility.cpp (hunk 9 above)
Comment (Yonas-net) [reply lib/base/utility.cpp:1317]: Because I want to cast it explicitly in a C++ way and not in `c-style`.

Comment (Al2Klimov) [reply lib/base/utility.cpp:1317]: Are **function style** casts **C style**?

**Comment by Yonas-net** [comment]
> If you just start Icinga 2 with an empty CRL configured, don't touch the CRL after start and establish your first connection, will Icinga 2 update the context?

So, if I don't configure the CRL path in `api.cnf` at all, no context update is performed. But if I configure that and the CRL does not contain any revoked certificates at all, the context is updated only once before connecting to an endpoint.

**Comment by Al2Klimov** [comment]
Only once? Where?

**Comment by Yonas-net** [comment]
> Only once? Where?

Well [here](https://github.com/Icinga/icinga2/pull/8515/files#diff-e9d85b2c0f02d6b34928833d4dc02612a3c2d85e9d13f1d9633a8e2feb8bd727R445)!

**Comment by Al2Klimov** [comment]
... **and** here:

https://github.com/Icinga/icinga2/blob/9b6326ce72d9e9d3156b987f8608019816bc993b/lib/remote/apilistener.cpp#L179

Do you consider your additional update neccessary?

**Comment by Yonas-net** [comment]
No, but I can't prevent it either!

**Comment by Al2Klimov** [comment]
You're writing this new code. Of course you can prevent it.

File: lib/remote/apilistener.cpp [hunk 10]
```diff
@@ -430,11 +430,31 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
 
 	auto& io (IoEngine::Get().GetIoContext());
 
+	auto lastModified = static_cast<time_t>(-1);
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:433]: `s/auto/time_t/` and drop the cast. (The C++ people say "**almost** always auto" for a reason.)

File: lib/remote/apilistener.cpp [hunk 11]
```diff
@@ -430,11 +430,31 @@ void ApiListener::ListenerCoroutineProc(boost::asio::yield_context yc, const Sha
...
+
+			if (!crlPath.IsEmpty()) {
+				time_t currentCreationTime = Utility::GetFileCreationTime(crlPath);
 
-			server->async_accept(sslConn->lowest_layer(), yc);
+				if (lastModified != currentCreationTime) {
+					UpdateSSLContext();
+				}
+
+				lastModified = currentCreationTime;
```
Comment (Al2Klimov) [inline lib/remote/apilistener.cpp:453]: Why is this one outside the above if?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
You've just addressed incoming connections here. What about outgoing ones?

**Comment by Yonas-net** [comment]
If the incoming connection has been checked and successfully validated at the master, why do I have to add an additional check for the outgoing one. We only need to update the master and not the ones connected to it, because this has no effect at all if the master doesn't notice anything.

**Comment by Al2Klimov** [comment]
Also address outgoing connections.

**Comment by Al2Klimov** [comment]
I don't see any actual changes.

**Comment by Yonas-net** [comment]
> Also address outgoing connections.

Master and agent disabled and I have revoked the certificates from the agent. After that I started both and master noticed that the crt is revoked. Agent turned off and undo crt revokation, agent started again, crt validation was successful and no error messages for revoked crts.

**Comment by Al2Klimov** [comment]
Fine. Undo the commit message change.

**Comment by Yonas-net** [comment]
Do you know why the Github actions fail?

**Comment by Al2Klimov** [comment]
Figure it out.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a test protocol.

**Comment by Yonas-net** [comment]
Testprotocol:

**Before**: See #8501

**After**:

Revoke the certificate while the master and the agent are stopped and then start them both.

Master Log:
```powershell
[2020-12-07 12:30:11 +0100] information/ApiListener: New client connection for identity 'ha-cluster' from [::1]:49781 (certificate validation failed: code 23: certificate revoked)
[2020-12-07 12:30:11 +0100] information/JsonRpcConnection: Received certificate request for CN 'ha-cluster' signed by our CA.
[2020-12-07 12:30:11 +0100] information/JsonRpcConnection: The certificate for CN 'ha-cluster' is valid and uptodate. Skipping automated renewal.
```

And now update the ``CRL`` in my case, I will undo the crt revocation.

```powershell
[2020-12-07 12:35:21 +0100] warning/JsonRpcConnection: API client disconnected for identity 'ha-cluster'
[2020-12-07 12:35:31 +0100] information/ApiListener: New client connection for identity 'ha-cluster' from [::1]:49823
[2020-12-07 12:35:31 +0100] information/ApiListener: Sending config updates for endpoint 'ha-cluster' in zone 'ha-cluster'.
```

**Review by Al2Klimov — APPROVED** [review]

**Comment by julianbrost** [comment]
Can you please test what happens in the following scenarios:
- Removing the CRL file completely
- Invalid CRL file (truncated file for example)

**Comment by Yonas-net** [comment]
> * Removing the CRL file completely

```powershell
[2020-12-07 12:56:28 +0100] critical/SSL: Error loading crl file '~/var/lib/icinga2/ca/ha-master.crl': 33558530, "error:02001002:system library:fopen:No such file or directory"
[2020-12-07 12:56:28 +0100] critical/config: Error: Cannot add certificate revocation list to SSL context for crl path: '~/var/lib/icinga2/ca/ha-master.crl'.
```

**Comment by julianbrost** [comment]
What is the general behavior in this case? Will Icinga terminate? If not, will if accept new connections? If it does, are these verified against the old CRL?

**Comment by Yonas-net** [comment]
So, if you enter CRL path, which does not exist at all, or the CRL file is invalid Icinga2 will terminate while compiling the ``api.conf`` file with this [error message](https://github.com/Icinga/icinga2/pull/8515#issuecomment-739873103).

**Review by julianbrost — CHANGES_REQUESTED** [review]
Please use `git rebase` to reorder your commits. Currently the first commit already uses `GetFileCreationTime` which is only introduced in the second one.

**Review by julianbrost — APPROVED** [review]
So as this generates a whole new context, this implicitly also reloads certificates and private keys, but only when a CRL exists and is changed. I think this is fine, as nothing bad should happen there unless someone manually messes around with stuff.

**Pull request merged**